File: plugins/modulation/phaser/phaser.js
// Phaser — 4-stage allpass cascade whose cutoff is swept by a sine LFO,
// then fed back into itself. Each allpass introduces a 90° phase shift
// at its cutoff; mixed back with the dry signal, the cumulative shift
// creates the deep notch-and-sweep characteristic of the MXR / Small
// Stone-style sound. Feedback sharpens the notches into resonant peaks.

function PhaserProcessor() {
  this.s = [0, 0, 0, 0];   // per-stage delay element
  this.fb = 0;
  this.phase = 0;
}
PhaserProcessor.parameterDescriptors = [
  { name: 'rate',     label: 'Rate (Hz)', defaultValue: 0.4, minValue: 0.05, maxValue: 4 },
  { name: 'depth',    label: 'Depth',     defaultValue: 0.8, minValue: 0,    maxValue: 1 },
  { name: 'feedback', label: 'Feedback',  defaultValue: 0.5, minValue: 0,    maxValue: 0.95 },
  { name: 'mix',      label: 'Mix',       defaultValue: 0.5, minValue: 0,    maxValue: 1 }
];
PhaserProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var rate = parameters.rate[0];
  var depth = parameters.depth[0];
  var feedback = parameters.feedback[0];
  var mix = parameters.mix[0];
  var phaseInc = (2 * Math.PI * rate) / sampleRate;
  for (var i = 0; i < input.length; i++) {
    var lfo = 0.5 + 0.5 * Math.sin(this.phase);
    // Sweep 250 .. 2500 Hz, scaled by depth.
    var cutoff = 250 + depth * 2250 * lfo;
    var w = Math.tan(Math.PI * cutoff / sampleRate);
    var a = (1 - w) / (1 + w);
    var x = input[i] + this.fb * feedback;
    for (var s = 0; s < 4; s++) {
      var y = a * x + this.s[s];
      this.s[s] = x - a * y;
      x = y;
    }
    this.fb = x;
    output[i] = input[i] * (1 - mix) + x * mix;
    this.phase += phaseInc;
    if (this.phase > 2 * Math.PI) this.phase -= 2 * Math.PI;
  }
  return true;
};

registerProcessor('phaser', PhaserProcessor);

File: plugins/modulation/phaser/Phaser.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Phaser (native port) — 4-stage allpass cascade swept by a sine LFO and
// fed back into itself. Each allpass introduces a 90° phase shift at
// cutoff; mixed back with the dry signal that produces the deep
// notch-and-sweep characteristic of MXR / Small Stone phasers. Feedback
// sharpens the notches into resonant peaks.
public final class Phaser implements VocalMonitorNativePlugin {
    private final float[] s = new float[4];
    private float fbState = 0f;
    private float phase = 0f;
    private int sampleRate = 44100;
    private float rate = 0.4f, depth = 0.8f, feedback = 0.5f, mix = 0.5f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        for (int i = 0; i < 4; i++) s[i] = 0f;
        fbState = 0f;
        phase = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "rate", "depth", "feedback", "mix" }; }
    @Override public float parameterMin(String n) {
        if ("rate".equals(n)) return 0.05f;
        return 0f;
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "rate":     return 4f;
            case "feedback": return 0.95f;
            default:         return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "rate":     return 0.4f;
            case "depth":    return 0.8f;
            case "feedback": return 0.5f;
            case "mix":      return 0.5f;
            default:         return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "rate":     return "Rate (Hz)";
            case "depth":    return "Depth";
            case "feedback": return "Feedback";
            case "mix":      return "Mix";
            default:         return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "rate":     rate = v; break;
            case "depth":    depth = v; break;
            case "feedback": feedback = v; break;
            case "mix":      mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float phaseInc = (float) (2.0 * Math.PI * rate / sampleRate);
        final float twoPi = (float) (2.0 * Math.PI);
        final float fb = feedback;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float depthLocal = depth;
        final float invSr = 1f / sampleRate;
        float ph = phase;
        float s0 = s[0], s1 = s[1], s2 = s[2], s3 = s[3];
        float fbS = fbState;

        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float lfo = 0.5f + 0.5f * (float) Math.sin(ph);
            float cutoff = 250f + depthLocal * 2250f * lfo;
            float w = (float) Math.tan(Math.PI * cutoff * invSr);
            float a = (1f - w) / (1f + w);

            float x = input[i] + fbS * fb;
            float y;
            y = a * x + s0; s0 = x - a * y; x = y;
            y = a * x + s1; s1 = x - a * y; x = y;
            y = a * x + s2; s2 = x - a * y; x = y;
            y = a * x + s3; s3 = x - a * y; x = y;
            fbS = x;
            output[i] = input[i] * dry + x * mixLocal;

            ph += phaseInc;
            if (ph > twoPi) ph -= twoPi;
        }
        s[0] = s0; s[1] = s1; s[2] = s2; s[3] = s3;
        fbState = fbS;
        phase = ph;
    }
}
